packages/field-base/src/slot-styles-mixin.js

```diff
@@ -0,0 +1,88 @@
+import { dedupingMixin } from '@polymer/polymer/lib/utils/mixin.js';
+
+const stylesMap = new WeakMap();
+
+/**
+ * Get all the styles inserted into root.
+ * @param {DocumentOrShadowRoot} root
+ * @return {Array<string>}
+ */
+function getRootStyles(root) {
+  if (!stylesMap.has(root)) {
+    stylesMap.set(root, []);
+  }
+
+  return stylesMap.get(root);
+}
+
+/**
+ * Detect if styles have been already inserted to root.
+ * @param {string} styles
+ * @param {DocumentOrShadowRoot} root
+ * @return {boolean}
+ */
+function hasStyles(styles, root) {
+  return getRootStyles(root).includes(styles);
+}
+
+/**
+ * Insert styles into the root.
+ * @param {string} css
+ * @param {DocumentOrShadowRoot} root
+ */
+function insertStyles(styles, root) {
+  if (!styles || (styles && hasStyles(styles, root))) {
+    return;
+  }
+
+  const rootNode = root === document ? document.head : root;
+  const style = document.createElement('style');
+  style.appendChild(document.createTextNode(styles));
+  rootNode.appendChild(style);
+}
+
+const SlotStylesMixinImplementation = (superclass) =>
+  class SlotStylesMixinClass extends superclass {
+    /**
+     * List of styles to insert into root.
+     * @protected
+     */
+    get slotStyles() {
+      return {};
+    }
+
+    /** @protected */
+    connectedCallback() {
+      super.connectedCallback();
+
+      this.__applySlotStyles();
```

Comment:
How to avoid naming collisions btw? If a number-field adds some styling to `input[slot='input']`, how do we avoid those styles from affecting the slotted input of a text-field? Since the styles are instance-specific, using an id in the selector could work.

Reply:
Good question. I think we could use `this.localName` to prefix styles and to produce something like

```css
vaadin-number-field > input[type='number'][slot='input'] {
  /* */
}
```

Reply:
Here is an example of how we can do it in `vaadin-number-field` to ensure `vaadin-integer-field` also works:

```js
/** @protected */
get slotStyles() {
  const tag = this.localName;
  return {
    number: `
      ${tag} input[type="number"]::-webkit-outer-spin-button,
      ${tag} input[type="number"]::-webkit-inner-spin-button {
        -webkit-appearance: none;
        margin: 0;
      }

      ${tag} input[type="number"] {
        -moz-appearance: textfield;
      }

      ${tag}[dir='rtl'] input[type="number"]::placeholder {
        direction: rtl;
      }

      ${tag}[dir='rtl']:not([has-controls]) input[type="number"]::placeholder {
        text-align: left;
      }
    `
  };
}
 ```